src/final1/subtests/ValidRecommendCommandTest.java

```diff
@@ -0,0 +1,261 @@
+	}
+
+	/**
+	 * Asserts correct results for different variations of spaces in the recommend command.
+	 */
+	@Test
+	public void spacesTest() {
+		String[] variants = {
+				"recommend UNION(S1 201,INTERSECTION(S1 105,S3 107))",
+				"recommend  UNION  (S1 201,INTERSECTION(S1 105,S3 107))",
+				"recommend UNION  (S1 201,INTERSECTION(S1 105,S3 107))",
+				"recommend UNION(S1  201,INTERSECTION(S1 105,S3 107))",
+				"recommend UNION(S1 201  ,INTERSECTION(S1 105,S3 107))",
+				"recommend UNION(S1 201,  INTERSECTION(S1 105,S3 107))",
+				"recommend UNION(S1 201,INTERSECTION  (S1 105,S3 107))",
+				"recommend UNION(S1 201,INTERSECTION(  S1 105,S3 107))",
+				"recommend UNION(S1 201,INTERSECTION(S1 105   ,S3 107))",
+				"recommend UNION(S1 201,INTERSECTION(S1   105,  S3 107))",
+				"recommend UNION(S1 201,INTERSECTION(S1 105,S3  107))",
+				"recommend UNION(S1 201,INTERSECTION(S1 105,S3 107  ))",
+				"recommend UNION(S1 201,INTERSECTION(S1 105,S3 107)  )",
+				"recommend UNION(S1 201,INTERSECTION(S1 105,S3 107))  ",
+				"recommend        UNION      (     S1     201    ,    INTERSECTION    (    S1    105   ,   S3    107    )    )     "
+		};
+		for (String variant : variants) {
+			oneLineTest(addQuit(variant), "calc:202,centos6:106,impress:203,libreoffice:200",
+				Input.getFile(TASK_SHEET_INPUT_FILE));
+		}
+	}
+
+	/**
+	 * Asserts overall correct behaviour of the implementation. This includes several error detection and recovery after
+	 * errors.
+	 */
+	@Test
+	public void oneLineTest() {
+		String[] queries = new String[] {
+				"recommend S1 1",
+				"recommend S1 3",
+				"recommend S4 1",
+				"recommend S2 1",
+				"recommend S2 2",
+				"redbutton",
+				"recommend S3 1",
+				"recommend S3 2"
+		};
+		// @formatter:off
+		List<Matcher<String>> matchers = getMatchers(
+			is(""),
+			startsWith("Error,"),
+			startsWith("Error,"),
+			is(""),
+			is("b:1"),
+			startsWith("Error,"),
+			is("a:2"),
+			is("")
+		);
+		// @formatter:on
+		// edges: new String[] { "b:1-[successor-of]->a", "a-[predecessor-of]->b" }
+		multiLineTest(addQuit(queries), matchers, Input.getFile(ONE_LINE_INPUT_FILE1));
+
+		queries = new String[] {
+				"recommend S1 1",
+				"recommend S1 2",
+				"recommend S2 2",
+				"recommend S3 2",
+		};
+		// @formatter:off
+        matchers = getMatchers(
+            startsWith("Error,"),
+            is(""),
+            is(""),
+            is("")
+        );
+        // @formatter:on
+		multiLineTest(addQuit(queries), matchers, Input.getFile(ONE_LINE_INPUT_FILE2));
+	}
+
+	/**
+	 * Asserts that product ID 0 can be handled correctly.
+	 */
+	@Test
+	public void zeroIdTest() {
+		multiLineTest(addQuit(new String[] {
+				"recommend S1 0",
+				"recommend S1 1",
+				"recommend S2 0"
+		}), new String[] {
+				"c:1",
+				"b:0",
+				""
+		}, Input.getFile(ZERO_ID_INPUT_FILE));
+	}
+
+	@Test
+	public void complexTest() {
+		String[][] commandResultArray = new String[][] {
+				{
```

Comment:
Probably not, I've originally thought through this on paper and then tested it with two different implementations.